### PR TITLE
Json schema

### PR DESCRIPTION
**Pull request by mucke2701**
JSON-Schema for KfWs Open Project Location Model. 
feature_project_schema.json <- Main Schema with geometries and project_data properties

master_project_schema.json <- Meta Schema for the project_core schema and the extensions schema like wdpa; referenced in feature_project_schema.json

wdpa_schema.json <- Extension on top of the project_core.schema for protected areas projects with extra attributes; referenced in master_project_schema

project_core_schema.json <- Project_core contains all attributes which are relevant for all projects; referenced in master_project_schema

dac5_schema.json <- schema for the dac5 purpose codes; referenced in project_core (extracted to schema for better oveview, maintenance in collaberation)

sector_location_schema.json <- schema for the sector and location_type pairs; referenced in project_core (extracted to schema for better oveview, maintenance in collaberation)

### CODE REVIEW

**Comment by fretchen** [comment]
Hi, 

very cool. I think that it would be good to integrate it into the existing stack before just merge. 

- I would propose to move the new schemas into the `references` folder in which we have already existing schemas.
- I would propose to remove the 1.0 folder. The version should best be done through tags. Why not use them ?
- I think that the explanations that have been added in the `tor/index.md` do not fit there. They should better be moved to the `docs/dev/index.md`, which needs to be updated anyways.
- Given the new structure of the schemas it would be quite helpful to update the documention in `docs/dev/index.md` and the automatically generated parts in `docs/dev/project_location_model.md`
- We have unit testing for the schemas and it should most likely be updated to deploy the new schemas to make sure that the excel etc is indeed compatible. Most importantly I think about the test `tests/test_validate_simple_schema.py`
- Is it correct that the `dac5_schema.json` replaces the existing `references/iati_sector_codes_v2_0_3.json`. Why is this replacement necessary ?
- Is it correct that the `project_core_schema.json` replaces the existing `references/project_location_schema.json`. Why is this replacement necessary and what are the most important changes ?
- Is it correct that the `feature_project_schema.json` extends the allowed types ? Is there any connection to existing columns in the [excel template](https://openkfw.github.io/open-geodata-model/technical_notes/#the-project-location-model) or  the [current model](https://openkfw.github.io/open-geodata-model/dev/project_location_schema/) ?
- I must admit that I cannot understand the logic of `master_project_schema`. What is its role compared to existing code ? It would seem that you have introduced some if else logic in there...
- What is the `WDPA_Schema` ? It sounds like a new feature that we did not have before.
- Are any of the proposed changes breaking changes that would require changes to the existing excel implementations etc ? Then we should clearly mark this, provide migration code etc... 

Taken together all these changes are quite substantial and I wonder if it would be better to split them into more manageable PRs...

**Review by fretchen — CHANGES_REQUESTED** [review]
See comment.

**Comment by mucke2701** [comment]
Thanks for the feedback. 
I will go ever the points in detail and adjust according to your suggestions hopefully next week.

**Comment by mucke2701** [comment]
- As requested I changed the directory of the schemas to the `references` folder.

- Versioning through tags is preferable so I've deleted the unneccessary folder hierarchy.

- The Documentation is currently in working progress by Fabi, you should see the commits in the history of this branch.

- I've added a test_validate_feature_schema.py which can be used to test future compatibilty and can be expanded. (2 Cases 1 Passes 1 Fails)

- I don't know if the dac5schema should substitute iati_sector_codes as there were no column in the Excel-Templates for iati_sector_codes only for dac5purpose codes. In the development of the schemas will solely focused on the Excel-Template and project-location-schema.json

- The feature_project_schema should substitute the project-location-schema. As it is better suited for geographic data structures. It will allow us to represent more complex geometries like Polygon etc. feature_project_schema is built hierarchially and is represented as Feature a known Geojson Format (Geometric objects with additional properties are Feature objects). In the properties is all the project information which is separated from the geometry. We reference the master_schema in the properties which for itself reference all different project types like core projects or wdpa projects and will apply the right schema conditionally. The approach of using multiple schemas which reference each other should help us to not bloat the project schemas with the categorical attributes like sector, location types or dac 5 purpose codes. Also we can use the feature_project_schema for all different project types in the future for validation. 

- We also added a new Excel-Template which is compatible with our Validator. But to validate Excel-Sheets we will always need some extra steps and logic which is out of scope of the schemes. Thats why we strive for a better suited data collection tool. 

- The Master Schema can be understand as Master Project Info Schema depending on the attributes given in a Json we can differentiate the project types. The Master Schema acts as a filter as it will apply the schema which is appriopriate for the project type. Right now we have the project type project_core but in near future there will be a project type for protected areas (wdpa_schema) where additional attributes are needed or some project_core attributes are no longer needed, depending which attributes are provided. That's why the logic is a bit hard to follow but usually should built future project types on top of the project_core attributes and have additionaly there specific attributes.

- WDPA Schema is for the protected areas projects I don't have much insight where and how there are currently used, but there was a request to start with the implementation of the schema. This schema is still in work. 

- I would say yes there are breaking changes but as far as I understood the schema which is currently in the Repo was never used to validate filled Excel-Sheets. We could provide a python for that as in `test_validate_simple_schema.py`. With the new attribute schemeVersion there is the need for our provided Excel Template. This Excel Template is also customized for our Validator which is not a part of this Repo.

- We can split this PR into multiple smaller ones, but I would prefer that we fine grind this PR as I would say all this changes are connected and the size of them should still be manageable. 

Merry Christmas and Happy New Year

**Comment by ffischer1984** [comment]
@fretchen 
> I would propose to remove the 1.0 folder. The version should best be done through tags. Why not use them ?

@mucke2701 
> Versioning through tags is preferable so I've deleted the unneccessary folder hierarchy.

Sorry guys I don't agree with you, here the arguments:
* if you use only git for versioning you've a tight coupling between the package-version and the schema version. Thats a bad behaviour:
* A tight coupling between this should be avoid bc it's impossible to bind two different package-versions of the same library in one software-product
* This leads to a big dependency-issue in our software-projects that we can only support the latest schema or have to deploy different validator-webpages for each schema version. This is a big disadvantage because it reduces our flexibility to the future developments. Thats the reason why I want the git-versioning process being decoupled by the schema-versions.

**Comment by ffischer1984** [comment]
@fretchen 
@mucke2701 
> I would propose to move the new schemas into the references folder in which we have already existing schemas.

If I understand it correctly, we're currently supporting two schemas:
* excel-schema which should be deprecated in the future (currently one possible 'import schema')
* current schema 'import/export schema' which should be the future schema

I would prefer to seperate the two schemas in different folder to improve the readability and to make it easier for newcomers to dig into it. How the folders are named I don't care,

**Comment by fretchen** [comment]
Hi,

this looks really massive now. Experience from previous PRs shows that it is much better for all of us if we separate this into smaller pieces.

It took my easily an hour or more to review this one and we are far from having it ready to merge. So let me try to propose things that are out of scope of this initial PR here:

- [ ] The WDPA should be separated into an extra PR. I started an issue #86 to follow up on it. This also means that I think that the logic around the master schema should be moved into the solution of #86.
- [ ] As for the command by @ffischer1984 on the path names I opened #87. Let us please move the discussion over there and keep it out of this one.

# Different schemas
- It would seem that you propose a major overwhole of the existing schemas. I cannot follow the logic without the documentation for the moment. 
- @ffischer1984  I do not know what you mean with two schemas. The excel template exists and there have been no public plans to depreceate it. The schema was meant as a reference for the template and other future implementations. I also do not know what you mean with the "current schema"

# Excel template
Please let us avoid breaking changes as much as possible. I had a look into the suggested changes for the excel. From what I can see it is an added column in the fill-me sheet. Is this true ? This column could be added into the sheet "version" as it is not really relevant for the user anyways. This way the fill-me sheet would not be changed and currently circulating excel files would keep their validity. This would avoid the need of writing migration routines for excel files from version 1.0 for all of us. 

# Tests

- Why did you remove the `test_iati_codes.py` and the `test_location_codes.py` ? What makes them depreceated ?
- If tests are remove them and let's keep the repo clean. However, I do not understand the reason why the existing tests should all be irrelevant by now.

Taken together, it sounds as if it was helpful to simplify this PR and maybe also reduce its ambition.

**Comment by mucke2701** [comment]
Hi, 

I will close this PR and will split the work progress in several PR's. One of our first targets is to use the JSON-Schemes as representation of the Open Geodata Model. 

So that from now on the project location model comes in form of an JSON Schema and a Excel-Template can be used to collect the data, but we are looking for better data collection tools.
